Add new `message_move_limit_seconds` realm setting.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First 4 commits are refactoring commits and to fix a live-update issue.
- Next commit adds the new `message_move_limit_seconds` setting.
- Next two commits are to add code to use the new setting to check topic edit permission.
- Next two commits are to add code to use the new setting to check stream edit permission.
- Last commit adds UI element for the new setting.

<!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Nice! We need to add an explanation to clarify that the time limit doesn't apply to admins and moderators. The parallel with the "Message deletion" section would be to add a line just under the "Moving messages" section heading, but it's a bit trickier, because it would be weird to say that moderators can move any message when they aren't allowed to move messages in the first place.

Maybe the best solution here is to put a note directly into the setting description. Something like this?

> Time limit for editing topics and moving messages _(applies only to members and guests)_

It might be good to make the "Message deletion" section work the same way, with a note on the time setting instead of at the top of the section. Perhaps:

> Time limit for deleting messages _(does not apply to administrators)_

sahil839 commented:
Added note to the setting labels.

![Screenshot from 2022-12-31 20-25-07](https://user-images.githubusercontent.com/35494118/210140835-9112efc9-6c34-4f2e-b981-29971c03b85a.png)

alya commented:
Thanks! I think we can drop the "Administrators can delete any message" line.

sahil839 commented:
I have kept that line since administrators can delete messages sent by others as well and the dropdown setting is only for the messages sent by user themselves. So, I think it's worth mentioning that administrators can delete any message (which means including the messages sent by others).

alya commented:
Ah, I see. Sure, that makes sense.

alya commented:
@timabbott please take a look at this one. As noted at https://github.com/zulip/zulip/pull/23906#issuecomment-1368989903, it's best to consider this a blocker for #23906.

timabbott commented on zerver/views/realm.py:
```diff
@@ -197,47 +197,32 @@ def update_realm(
-            )
+
+        if setting_value_changed:
             data["message_content_delete_limit_seconds"] = message_content_delete_limit_seconds
```
Just a note that I think this return value code is planned to be deleted in favor of the `ignored_parameters_unsupported` system that @laurynmm was working on as a better way to communicate over the API which of the parts of a request we in fact processed -- but we can do that refactor when it's appropriate.

timabbott commented on zerver/models.py:
```diff
@@ -379,6 +379,12 @@ class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stub
 
     DEFAULT_COMMUNITY_TOPIC_EDITING_LIMIT_SECONDS = 259200
 
+    DEFAULT_MOVE_MESSAGE_LIMIT_SECONDS = 604800
```
We should spell the math out so it's clear this is 7 days -- 86400 * 7

timabbott replied:
It might even be worth having a constant at the time of the file `SECONDS_PER_DAY = 86400` and have logic like this use it.

sahil839 replied:
Done.

timabbott commented on zerver/tests/test_message_edit.py:
```diff
@@ -2148,6 +2148,89 @@ def check_move_message_according_to_policy(role: int, expect_fail: bool = False)
+            user: UserProfile,
+            old_stream: Stream,
+            new_stream: Stream,
+            error_msg: Optional[str] = None,
```
I would call this `expect_error_message` and make it a mandatory kwarg; doing so makes the code calling this function much more readable.

sahil839 replied:
Done.

timabbott commented on static/templates/settings/organization_permissions_admin.hbs:
```diff
@@ -191,18 +191,37 @@
+                      name="realm_message_move_limit_minutes"
+                      class="time-limit-custom-input"
+                      autocomplete="off"
+                      value="{{ realm_message_move_limit_seconds }}"/>
```
How does the seconds/minutes conversion work here? It seems like the value displayed here is a number of seconds, even though the input field name is a number of minutes.

sahil839 replied:
I have removed this line and we instead set the value in JS since we already set the dropdown value and visibility of custom input in JS. It also helps in live-update code.

timabbott commented:
Overall, this looks pretty good. I posted a few inline comments on things I didn't understand. The main question I have is whether it makes sense to have two different time limit settings -- one for topics within a stream and one for moving topics to a different stream. The reasoning here is that I can imagine someone wanting the flexibility, and I think it'd be a pain to migrate this to that after merging it, but my read of this PR is that it'd be relatively easy to just offer two settings. Idunno what the best names would be; maybe `move_messages_between_streams_time_limit` and `edit_topic_time_limit` would be most parallel to the existing `_policy` setting names, but I could also see trying to make these parallel to each other (with a view that we'd later rename `edit_topic_policy`). 

Anyway, the benefit is that we'd then just have three cleanly independent sets of settings -- those for editing content, those for editing topics and those for editing streams, and not need special JS logic for disabling/enabling inputs based on the values of multiple inputs, etc -- it's just feel more like a consistent UI framework. 

@alya agreed that's probably a good plan if it's a relatively small effort to rework this to that model (such that even though it's not actually a priority for these to be independent, the fact that we'd save an annoying migration if/when we want to do that in a year or two would be worth the effort to adjust this now).

> Time limit for editing topics and moving messages (applies only to members and guests)

Can we write this as "(does not apply to moderators and administrators)"? (Long term, I imagine we'll have a setting that is "What groups are exempt from the time limit?", with a default of "moderators")

sahil839 commented:
This is still not ready for review. There a couple of live-update issues that I need to look at.

sahil839 commented:
Updated the PR. Posted replies to the comments above as well. Writing the major changes below.
- Added two different settings for topic and stream edits and have named them `move_messages_within_stream_limit_seconds` and `move_messages_between_streams_limit_seconds` to keep it somewhat consistent with settings for edit and delete limit and also did not use `edit_topic_policy` for topic edit setting as Tim suggested above.
- Replaced `(applies only to members and guests)` with `(does not apply to moderators and administrators)`.
- Added commit to live-update the pencil icon in recipient row.

sahil839 commented on frontend_tests/node_tests/settings_org.js:
```diff
@@ -542,6 +548,18 @@ test("set_up", ({override, override_rewire}) => {
+        "<stub move_between_streams_limit parent",
+    );
+    $("#id_realm_move_messages_between_streams_limit_seconds").set_parent(
+        $stub_move_between_streams_limit_parent,
```
There is too much repetition of this code. I think we can either mock the function which uses these elements and `find` and `parent` functions or we can extract a function and call it for different settings.

timabbott commented:
I merged the first 4 commits as the series ending with 5b1f6a696e60f4f618561ff462d7ff1941e6927e, just to shrink this a bit.

timabbott commented on static/js/message_edit.js:
```diff
@@ -305,6 +306,19 @@ export function end_if_focused_on_message_row_edit() {
+        // There is no need to rerender if user is already not
+        // allowed to edit topics as per "edit_topic_policy".
+        return;
+    }
```
Hmm, what happens if the user previously had that permission? I'll drop this hunk for today's merge, we can do a refinement in a follow-up. (I could imagine checking if we had the permission either before or after).

sahil839 replied:
So, this condition only checked whether the user had permission to edit topic as per role-based `edit_topic_policy` setting and we call this function only on updating the time-limit setting and not when updating `edit_topic_policy`.

timabbott commented:
I pushed a few documentation changes and removed the optimization in https://github.com/zulip/zulip/pull/23916#discussion_r1100621853 since I didn't understand it; with those changes this is good to merge.

@alya can you do a documentation update to go with this today?
